pkg/auth/providers/saml/saml_provider_test.go

```diff
@@ -0,0 +1,132 @@
+package saml
+
+import (
+	"context"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/providers/ldap"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"testing"
+)
+
+func createSamlProviderWithMockedLdap(name string, hasLdap bool) *Provider {
+	provider := &Provider{name: name}
+	provider.ldapProvider = &mockLdapProvider{isLdapConfigured: hasLdap, providerName: provider.name}
+	return provider
+}
+
+func TestConfiguredOktaProviderContainsLdapProvider(t *testing.T) {
+	// saml.Configure runs some ldap specific logic based on the saml provider name, so we provide
+	// just enough scaffolding to run the Configure function.
+	ctx := context.Background()
+	mgmtCtx, err := config.NewScaledContext(rest.Config{}, nil)
+	require.NoError(t, err, "Failed to create NewScaledContext")
+	tokenMGR := tokens.NewManager(ctx, mgmtCtx)
+	provider, ok := Configure(ctx, mgmtCtx, mgmtCtx.UserManager, tokenMGR, "okta").(*Provider)
+	require.True(t, ok, "Failed to Configure a valid Provider")
+
+	assert.True(t, provider.hasLdapGroupSearch(), "Missing LDAP group search capability for okta provider")
+	assert.NotNil(t, provider.ldapProvider, "Configured okta provider did not receive child LDAP provider")
+}
+
+func TestSearchPrincipals(t *testing.T) {
+	var userSearchKey = "al"
+	// Note: The mocked ldap provider alawys returns a single user named "alice"
+	testcases := []struct {
+		name              string
+		providerName      string
+		expectedLoginName string
+		isLdapConfigured  bool
+	}{
+		{
+			name:              "okta with ldap provides ldap user",
+			providerName:      "okta",
+			expectedLoginName: "alice",
+			isLdapConfigured:  true,
+		},
+		{
+			name:              "okta without ldap uses fallback behavior",
+			providerName:      "okta",
+			expectedLoginName: "al",
+			isLdapConfigured:  false,
+		},
+	}
+
+	for _, tt := range testcases {
+		t.Run(tt.name, func(t *testing.T) {
+			provider := createSamlProviderWithMockedLdap(tt.providerName, tt.isLdapConfigured)
+			results, err := provider.SearchPrincipals(userSearchKey, "user", v3.Token{})
```

Comment:
Is this test really only testing the mocked SearchPrincipals function rather than any code in the actual saml_provider?

Reply:
Oh, I think I see what you're going for here.